File: day07/part2.js
let total = 0;
for await (const line of console) {
    const pieces = line.split(": ");
    const target = Number(pieces[0]);
    const [first, ...numbers] = pieces[1].split(" ").map((n) => Number(n));

    if (matches(target, numbers, first)) {
        total += target;
    }
}

console.log(total);

function matches(target, numbers, currentValue) {
    if (numbers.length === 0) {
        return currentValue === target;
    }

    const [val, ...rest] = numbers

    return matches(target, rest, currentValue + val)
        || matches(target, rest, currentValue * val)
        || matches(target, rest, Number(`${currentValue}${val}`));
}

File: day07/part2.swift
var total = 0
while let line = readLine() {
    let pieces = line.split(separator: ": ")
    let target = Int(pieces[0])!
    let numbers = pieces[1].split(separator: " ").map { Int($0)!}

    if (matches(target: target, numbers: numbers, i: 1, currentValue: numbers[0])) {
        total += target
    }
}

print(total)

func matches(target: Int, numbers: [Int], i: Int, currentValue: Int) -> Bool {
    if (i == numbers.count) {
        return currentValue == target
    }

    let num = numbers[i]

    return matches(target: target, numbers: numbers, i: i + 1, currentValue: currentValue + num)
        || matches(target: target, numbers: numbers, i: i + 1, currentValue: currentValue * num)
        || matches(target: target, numbers: numbers, i: i + 1, currentValue: Int("\(currentValue)\(num)")!)
}